File: Step-3(Arrays)/Step 3.3 Arrays(Hard)/493_Reverse Pairs.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var merge = (a, low, mid, high) => {
  let left = low,
    right = mid + 1;
  let temp = [];

  while (left <= mid && right <= high) {
    if (a[left] <= a[right]) temp.push(a[left++]);
    else temp.push(a[right++]);
  }
  while (left <= mid) temp.push(a[left++]);
  while (right <= high) temp.push(a[right++]);

  for (let i = low; i <= high; i++) a[i] = temp[i - low];
};
var countReversePairs = (nums, low, mid, high) => {
  let count = 0;
  let right = mid + 1;

  for (let i = low; i <= mid; i++) {
    while (right <= high && nums[i] > 2 * nums[right]) right++;
    count += right - (mid + 1);
  }

  return count;
};

var mergeSort = (a, low, high) => {
  let cnt = 0;
  if (low >= high) return cnt;
  let mid = Math.floor((low + high) / 2);
  cnt += mergeSort(a, low, mid);
  cnt += mergeSort(a, mid + 1, high);
  cnt += countReversePairs(a, low, mid, high);
  merge(a, low, mid, high);

  return cnt;
};

var reversePairs = (nums) => {
  return mergeSort(nums, 0, nums.length - 1);
};

File: Step-3(Arrays)/Step 3.3 Arrays(Hard)/493_Reverse Pairs.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
  void merge(vector<int> &nums, int low, int mid, int high)
  {
    vector<int> temp;
    int left = low;
    int right = mid + 1;

    while (left <= mid && right <= high)
    {
      if (nums[left] <= nums[right])
      {
        temp.push_back(nums[left]);
        left++;
      }
      else
      {
        temp.push_back(nums[right]);
        right++;
      }
    }
    while (left <= mid)
    {
      temp.push_back(nums[left]);
      left++;
    }
    while (right <= high)
    {
      temp.push_back(nums[right]);
      right++;
    }

    for (int i = low; i <= high; i++)
    {
      nums[i] = temp[i - low];
    }
  }

  int countReversePairs(vector<int> &nums, int low, int mid, int high)
  {
    int count = 0;
    int right = mid + 1;

    for (int i = low; i <= mid; i++)
    {
      while (right <= high && nums[i] > 2 * (long)(nums[right]))
        right++;
      count += (right - (mid + 1));
    }

    return count;
  }

  int merge_Sort(vector<int> &nums, int low, int high)
  {
    int count = 0;
    if (low >= high)
      return count;

    int mid = (low + high) / 2;

    count += merge_Sort(nums, low, mid);
    count += merge_Sort(nums, mid + 1, high);
    count += countReversePairs(nums, low, mid, high);

    merge(nums, low, mid, high);

    return count;
  }

public:
  int reversePairs(vector<int> &nums)
  {
    int n = nums.size();
    return merge_Sort(nums, 0, n - 1);
  }
};

int main()
{
  return 0;
}